### PR TITLE
v1.0.0 Release - Production-Ready PathFinder

### PR DESCRIPTION
**Pull request by J-Yellen**
# Pull Request: v1.0.0 Release - Production-Ready PathFinder

## Overview

This PR merges the `release/v1.0.0` branch into `main` for the **v1.0.0 release**, marking PathFinder as production-ready with comprehensive enhancements, critical bug fixes, and significantly improved test coverage.

## Key Metrics

- **Test Coverage**: 60% → 95% (+35%)
- **Test Count**: 19 → 105 tests (+86 tests)
- **Critical Bugs Fixed**: 2 major issues resolved
- **New Features**: Enhanced visualization, simplified plot API, convenience API
- **Documentation**: Complete API reference, examples, and migration guides

---

#### Simplified API - `bam` Parameter Optional
**Major usability improvement**: The `bam` parameter is now optional and automatically extracted from HDFS/WHDFS objects:

```python
# New simplified usage - just pass results object
fig, ax = plot_results.plot(whdfs, highlight_top_path=True, size=12)

# Still backward compatible - explicit bam works
fig, ax = plot_results.plot(bam, whdfs, size=12)
```

**Benefits**:
- Cleaner API following DRY principle (don't repeat data)
- More intuitive: `plot(whdfs)` instead of `plot(bam, whdfs)`
- Flexible parameter order: first argument can be Results object
- Maintains full backward compatibility

#### `highlight_top_path` Parameter
Visually emphasise the best path in Binary Acceptance Matrix plots:
- Semi-transparent green overlay on rows corresponding to top path
- Highlighting stops at diagonal (lower triangle only)
- Axis labels for top path features colored dark green and bold

```python
# Works with simplified API
fig, ax = plot_results.plot(whdfs, highlight_top_path=True, size=12)
```

#### `axis_labels` Parameter  
Automatic label detection from BinaryAcceptance objects:
- Uses `bam.labels` when `axis_labels=True`
- Respects BAM sorting order (works with WHDFS auto_sort)
- `xy_labels` parameter still works as override

```python
bam = pf.BinaryAcceptance(matrix, weights, labels=['A', 'B', 'C'])
whdfs = pf.WHDFS(bam, top=5)
whdfs.find_paths()

# Simplified API with automatic labels
fig, ax = plot_results.plot(whdfs, axis_labels=True)
```

### 2. High-Level API (`find_best_combinations`)

New convenience function wrapping the complete PathFinder workflow:

```python
results = pf.find_best_combinations(
    matrix=correlation,
    weights=weights,
    threshold=0.7,
    top=10,
    labels=feature_names
)
# Returns Results with automatic remapping - ready to use!
```

#### `get_sorted_results()` Method
Access results in sorted index space for debugging/analysis:

```python
whdfs = pf.WHDFS(bam, top=5, auto_sort=True)
whdfs.find_paths()
sorted_results = whdfs.get_sorted_results()  # Sorted space
original_results = whdfs.get_paths  # Original space (automatic)
```

#### `get_sorted_paths()` Method
Get paths in sorted index space:

```python
sorted_paths = whdfs.get_sorted_paths()  # Match sorted BAM
```

#### Enhanced `__str__()` 
String representation automatically shows remapped paths in original indices for user-friendly output.

### 4. Results Class Enhancements

New classmethod constructors for flexible object creation:

```python
# From list of Result objects
results = Results.from_list_of_results(result_list)

# From list of dictionaries
results = Results.from_list_of_dicts(dict_list)
```

### 5. Flexible Configuration

- `top=None` support for unlimited path collection in HDFS/WHDFS/Results
- Optional `index_map` parameter in `remap_path()` (auto-detects from BAM)
- `plot_sorted` parameter for visualising sorted vs original index space

---

### Bug #1: WHDFS Initialization Error (CRITICAL)

**Issue**: `WHDFS.__init__` incorrectly set `self.top_weight = self._top_weights_default`, shadowing the `top_weight()` method definition.

**Impact**: 
- `set_top_weight()` didn't work
- Custom weight threshold functions couldn't be applied
- Affected algorithm pruning behavior

**Fix**: Changed to `self._top_weights = self._top_weights_default`

**File**: `src/pathfinder/dfs.py` line 180

**Test Coverage**: `test_whdfs_set_top_weight()` validates fix

### Bug #2: Plotting Path Alignment (CRITICAL)

**Issue**: When WHDFS used `auto_sort=True`, plots displayed the sorted BAM matrix but overlaid paths in original index space, causing paths to appear on invalid edges (black squares).

**Impact**:
- Visual misrepresentation of results
- Paths appeared to violate constraints
- Difficult to interpret results

**Fix**: 
- Detect sorted BAM via `_index_map` 
- When `plot_sorted=False`, unsort matrix for display
- Ensures paths (original indices) align with displayed matrix

**File**: `src/pathfinder/plot_results.py`

**Test Coverage**: `test_paths_dont_land_on_black_squares()` prevents regression

---

#### `dfs.py` Coverage (+18 tests)
- `get_sorted_results()` and `get_sorted_paths()` methods
- String representation with/without auto_sort remapping  
- Property setters: `weight_func`, `wlimit_func`, `set_top_weight`
- Default weight threshold behavior
- Static method `HDFS.chunked()`
- Path pruning and traversal edge cases
- Index remapping with/without explicit index_map
- Type conversions and target node identification
- Weight offset handling
- Visited node handling

#### `plot_results.py` Coverage (+6 tests)
- Simplified API: `plot(results)` extracts BAM automatically
- Parameter validation when BAM cannot be extracted
- Backward compatibility with explicit `bam` parameter
- Empty results handling with `highlight_top_path`
- `axis_labels` behavior when BAM has no labels
- Combined `highlight_top_path` and `axis_labels` functionality
- Single-node path highlighting

#### Critical Validation Tests
- `test_paths_dont_land_on_black_squares()`: Validates paths only traverse valid edges
- `test_plot_sorted_parameter()`: Verifies index space handling
- Edge case coverage for fully connected/disconnected graphs
- Boundary condition testing (minimal graphs, threshold limits)

### Coverage Metrics

| Module | Before | After | Change |
|--------|--------|-------|--------|
| Overall | 60% | 95% | +35% |
| `dfs.py` | ~70% | ~94% | +24% |
| `plot_results.py` | ~75% | ~93% | +18% |
| `result.py` | ~80% | ~99% | +19% |
| **Total Tests** | **19** | **105** | **+86** |

---

### README.md
- Updated badges: 105 tests passing, 95% coverage
- Enhanced Example 3 with simplified plot API usage
- New comprehensive API reference for `plot_results.plot()`
- Documented simplified API: `plot(whdfs)` instead of `plot(bam, whdfs)`
- Documented all new parameters with examples
- Updated testing section with current metrics

### CHANGELOG.md  
- Complete v1.0.0 release notes with feature descriptions
- Migration guide from 0.2.x to 1.0.0
- Detailed bug fix explanations
- Updated test coverage statistics
- Release highlights and stability commitment

### Tutorial Notebook (`path_test.ipynb`)
- Updated examples to use `highlight_top_path=True`
- Demonstrates new visualisation features
- Shows `find_best_combinations()` convenience function
- Updated for current best practices

---

### Breaking Changes
**None** - Full backward compatibility maintained

### Deprecated
**None**

### New Public API
- `plot(bam=None, results=None, ...)` - simplified API with optional bam parameter
- `plot(..., highlight_top_path=False, axis_labels=True)` - enhanced visualisation
- `find_best_combinations(matrix, weights, threshold, ...)` - convenience function
- `WHDFS.get_sorted_results()` - access sorted index space
- `WHDFS.get_sorted_paths()` - get paths in sorted indices
- `Results.from_list_of_results(result_list)` - classmethod constructor
- `Results.from_list_of_dicts(dict_list)` - classmethod constructor

---

## Checklist

- [x] All tests passing (103/103)
- [x] Test coverage ≥95%
- [x] Documentation updated (README, CHANGELOG)
- [x] No breaking changes
- [x] Critical bugs fixed
- [x] New features documented with examples
- [x] Tutorial notebook updated
- [x] Migration guide provided
- [x] Code follows project conventions (British English, type hints)
- [x] Git history clean and organized

---

### Run Full Test Suite
```bash
cd /Users/jamieyellen/Projects/PathFinder
pytest
```

**Expected**: All 105 tests pass

### Check Coverage
```bash
pytest --cov=src/pathfinder --cov-report=term-missing
```

**Expected**: ≥95% coverage

#### Test Visualisation Enhancements
```python
import pathfinder as pf
import numpy as np
from pathfinder import plot_results

# Create test data
n = 10
matrix = np.random.rand(n, n)
matrix = (matrix + matrix.T) / 2
weights = np.random.rand(n)
labels = [f'Feature_{i}' for i in range(n)]

# Test new parameters
bam = pf.BinaryAcceptance(matrix, weights, labels=labels, threshold=0.5)
whdfs = pf.WHDFS(bam, top=5)
whdfs.find_paths()

# Test simplified API - just pass results object
fig, ax = plot_results.plot(whdfs, 
                            highlight_top_path=True,
                            axis_labels=True,
                            size=10)

# Verify backward compatibility
fig2, ax2 = plot_results.plot(bam, whdfs, highlight_top_path=True)
```

#### Test Convenience Function
```python
results = pf.find_best_combinations(
    matrix=matrix,
    weights=weights,
    threshold=0.5,
    top=5,
    labels=labels
)
print(f"Best path: {results.get_paths[0]}")
print(f"Weight: {results.get_weights[0]:.3f}")
```

#### Test Bug Fixes
```python
# WHDFS set_top_weight should work
whdfs = pf.WHDFS(bam, top=5)
def custom_weight():
    return 10.0
whdfs.set_top_weight(custom_weight)
assert whdfs.top_weight() == 10.0

# Paths should not land on black squares
# (validated by test suite)
```

---

### Version Bump
The version will be automatically set to `1.0.0` via `setuptools_scm` after merging to main and tagging:

```bash
git checkout main
git merge release/v1.0.0
git tag -a v1.0.0 -m "Release v1.0.0: Production-ready PathFinder"
git push origin main --tags
```

### Post-Merge Steps
1. Verify version: `python -c "import pathfinder; print(pathfinder.__version__)"`
2. Build distribution: `python -m build`
3. Test installation: `pip install dist/pathfinder-1.0.0*.whl`
4. Create GitHub release with CHANGELOG excerpt
5. (Optional) Publish to PyPI: `python -m twine upload dist/*`

---

## Impact Summary

This release transforms PathFinder from a research tool to a production-ready library suitable for:

- **Machine Learning Pipelines**: Feature selection with pairwise constraints
- **Scientific Computing**: Signal region combination in physics
- **Data Science**: Correlation-aware variable selection
- **Production Systems**: Stable API, comprehensive testing, clear documentation

**Recommendation**: Approve and merge to establish v1.0.0 as the stable baseline for future development.

---

## Reviewers

Please focus review on:
1. Critical bug fixes (WHDFS initialization, plotting alignment)
2. New API stability and design
3. Test coverage comprehensiveness  
4. Documentation clarity and completeness

---

## Additional Notes

- All commits follow conventional commit format
- British English spelling maintained throughout
- Type hints and docstrings complete
- No external dependency changes
- Backward compatible with 0.2.x

**Merge Strategy**: Squash and merge or standard merge (both acceptable)

### CODE REVIEW

File: tests/test_dfs.py [hunk 1]
```diff
@@ -73,10 +82,593 @@ def test_basic_matrix2():
...
+    initial_count = len(whdfs.get_paths)
+    # Run again without resetting results
+    whdfs.find_paths(reset_result=False, verbose=False)
+    # Results should accumulate (or at least not reset)
+    assert len(whdfs.get_paths) >= initial_count
+
+
+def test_convenience_function_with_hdfs():
+    """Test find_best_combinations with HDFS algorithm"""
+    import pathfinder as pf
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_dfs.py]: removed unnecessary import

File: tests/test_dfs.py [hunk 2]
```diff
@@ -73,10 +82,593 @@ def test_basic_matrix2():
...
+        algorithm='hdfs',
+        verbose=False
+    )
+    assert len(results.get_paths) > 0
+
+
+def test_convenience_function_invalid_algorithm():
+    """Test find_best_combinations with invalid algorithm"""
+    import pytest
+    import pathfinder as pf
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_dfs.py]: removed unnecessary import

File: tests/test_dfs.py [hunk 3]
```diff
@@ -73,10 +82,593 @@ def test_basic_matrix2():
...
+    with pytest.raises(ValueError, match="Unknown algorithm"):
+        pf.find_best_combinations(
+            matrix=pseudo,
+            algorithm='invalid_algo'
+        )
+
+
+def test_convenience_function_with_runs():
+    """Test find_best_combinations with runs parameter"""
+    import pathfinder as pf
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_dfs.py]: removed unnecessary import

File: tests/test_dfs.py [hunk 4]
```diff
@@ -73,10 +82,593 @@ def test_basic_matrix2():
...
+        top=2,
+        runs=2,
+        verbose=False
+    )
+    assert len(results.get_paths) > 0
+
+
+def test_convenience_function_with_labels():
+    """Test find_best_combinations with labels parameter"""
+    import pathfinder as pf
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_dfs.py]: removed unnecessary import

File: tests/test_dfs.py [hunk 5]
```diff
@@ -73,10 +82,593 @@ def test_basic_matrix2():
...
+    bam = BinaryAcceptance(pseudo, weights=weights)
+    whdfs = WHDFS(binary_acceptance_obj=bam, top=1, allow_subset=False)
+    whdfs.find_paths(verbose=False)
+    # Should find at least one path
+    assert len(whdfs.get_paths) > 0
+
+
+def test_version_import():
+    """Test that __version__ is accessible"""
+    import pathfinder as pf
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_dfs.py]: removed unnecessary import

File: src/pathfinder/dfs.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+from itertools import islice
+import numpy as np
+from numpy import ndarray
+from .matrix_handler import BinaryAcceptance
+from .result import Results
+from typing import Iterable, Iterator, List, Optional, Callable, Union, Dict
+
+
+class HDFS(Results):
+
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply src/pathfinder/dfs.py:16]: removed cross object comparison and ensured like-for-like only

File: src/pathfinder/dfs.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+        Returns:
+            Results object with remapped paths.
+        """
+        if index_map is None:
+            index_map = self.bam._index_map
+        return super().remap_path(index_map, weight_offset)
+
+
+class WHDFS(Results):
+
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply src/pathfinder/dfs.py:179]: removed cross object comparison and ensured like-for-like only

File: tests/test_result.py [hunk 8]
```diff
@@ -71,13 +72,203 @@ def test_bulk_add_res():
...
     assert map_result1 == map_result2
-    assert map_result1[1]['Path'] == {map1[i] for i in paths[1]}
+    assert map_result1.get_paths[1] == sorted([map1[i] for i in paths[1]])
+    assert map_result1.get_raw_paths[1] == {map1[i] for i in paths[1]}
+
+
+def test_results_to_json_to_file():
+    """Test Results JSON serialization to file"""
+    import tempfile
+    import os
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_result.py]: moved imports

File: tests/test_result.py [hunk 9]
```diff
@@ -71,13 +72,203 @@ def test_bulk_add_res():
...
+    assert len(results.get_paths) == 1
+    assert results.get_paths[0] == []
+    assert results.best is not None
+
+
+def test_results_mismatched_lengths():
+    """Test Results raises error when paths and weights lengths don't match"""
+    import pytest
+    with pytest.raises(ValueError, match="Unequal length"):
+        Results([{0}], [1.0, 2.0], top=1)
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_result.py]: moved imports

File: src/pathfinder/result.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+        return lo_
+
+    def res_sort(self, trim: bool = True) -> None:
+        """
+        Sort the internal results list by weight in descending order.
+        Arguments:
+            trim (bool, optional): Trim the results list to the top N results. Defaults to
+        """
+        if trim:
+            self._res = sorted(self._res, reverse=True)[:self._top]
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply src/pathfinder/result.py:237]: Added an else statement and new tests

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+Repository = "https://github.com/J-Yellen/PathFinder"
+
+[project.optional-dependencies]
+dev = [
+    "pytest>=7.0.0",
+    "pytest-cov>=4.0.0",
+]
+
+[tool.pytest.ini_options]
+addopts = "--strict-markers --cov=pathfinder --cov-report=term-missing --cov-fail-under=90"
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply pyproject.toml:39]: no issue here the minimum coverage is 90% the code coverage is currently at 95%

File: tests/test_equality.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+    matrix = np.array([[True, False, True],
+                       [False, True, False],
+                       [True, False, True]])
+
+    bam = BinaryAcceptance(matrix, weights=[1.0, 2.0, 3.0])
+    hdfs1 = HDFS(bam, top=5)
+    hdfs2 = HDFS(bam, top=5)
+
+    # Change weight function
+    hdfs2.weight_func = lambda x: sum(x)
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_equality.py]: Now using sum (the dangers of automated test writing)

File: tests/test_equality.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+    matrix = np.array([[True, False, True],
+                       [False, True, False],
+                       [True, False, True]])
+
+    bam = BinaryAcceptance(matrix, weights=[1.0, 2.0, 3.0])
+    whdfs1 = WHDFS(bam, top=5, auto_sort=False)
+    whdfs2 = WHDFS(bam, top=5, auto_sort=False)
+
+    # Change wlimit function
+    whdfs2.wlimit_func = lambda x: sum(x)
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply tests/test_equality.py]: Now using sum (the dangers of automated test writing)

File: src/pathfinder/plot_results.py [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+    return ret
+
+
+def add_results(
+        ax: Axes,
+        res: Results,
+        lim: int,
+        plot_sorted: bool = False,
+        bam: Optional[BinaryAcceptance] = None
+) -> None:
```
[the remark this answers is not in the document]
Comment (J-Yellen) [reply src/pathfinder/plot_results.py:149]: removed unused bam parameter
